### PR TITLE
Split cram sanitation from public library

### PR DESCRIPTION
**Pull request by rgrinberg**
In the current setup, dune's output sanitation logic leaks into cram.
This is not desirable because we are planning to make cram a public
library.

This PR attempts to solve this problem by introducing a tiny cram
library where the user can craft his own cram binary and set the
sanitizer however he wishes.

The advantage is that the sanitizer can be composed with sanitizer
already defined by cram, and that everything is done in one process. The
cost is a tiny library introduced.

### CODE REVIEW

**Comment by jeremiedimino** [comment]
Just to confirm, the intended usage is to define a new cram binary and do:

```
(rule
 (alias runtest)
 (action (run ./mycram.exe run run.t)))
```

File: otherlibs/cram/src/dune_cramlib.mli [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    val make : (string -> string) -> t
+
+    val rewrite_build_path_prefix_map : t
+
+    module O : sig
+      val ( >>> ) : t -> t -> t
+    end
+  end
+
+  type t
```
Comment (jeremiedimino) [inline otherlibs/cram/src/dune_cramlib.mli:18]: Why do we need this type?

File: otherlibs/cram/src/dune_cramlib.mli [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    val default : t
+
+    val id : t
+
+    val make : (string -> string) -> t
+
+    val rewrite_build_path_prefix_map : t
+
+    module O : sig
+      val ( >>> ) : t -> t -> t
```
Comment (jeremiedimino) [inline otherlibs/cram/src/dune_cramlib.mli:14]: Not sure an operator is justified here. What about just:

```ocaml
val concat : t list -> t
```

?

**Comment by jeremiedimino** [comment]
I'll be honest, I'm not a huge fan of the "build your own tool" concept. It feels to me that the tool is not finished and you have to finish it yourself. Though, here it's not the most important.

**Comment by rgrinberg** [comment]
> I'll be honest, I'm not a huge fan of the "build your own tool" concept. It feels to me that the tool is not finished and you have to finish it yourself. Though, here it's not the most important.

This PR doesn't do this yet, but there will be a "standard" `dune-cram` binary that will be installed. The library is only necessary for those who are using their own sanitizers.

A bit of motivation why I'm doing this: I'd like to use dune-cram to write some tests for LSP and I've realized that I will need my own sanitizer. It seems like the sanitizer use case is common enough that we should support it properly.

**Comment by jeremiedimino** [comment]
Agreed that we should support it. I just wish we could just rely on the `--sanitizer` option. i.e. document the expected behaviour of the command passed to `--sanitizer` and let people pass their own tool that wouldn't need any special library.

The only annoying thing right now is that the sanitizer has a weird behaviour and adds these two whitespaces which is odd to document.

**Comment by rgrinberg** [comment]
> The only annoying thing right now is that the sanitizer has a weird behaviour and adds these two whitespaces which is odd to document.

There's another disadvantage. If the user wants to write their own sanitizer but preserve filtering according to `BUILD_PATH_PREFIX_MAP` then they must reimplement the logic.

I understand the desire to keep the interface tight, but not at the expense of making the tool more useful to users.

**Comment by jeremiedimino** [comment]
As discussed yesterday, let's merge the tool with the dune binary and let a `cram` stanza be the only visible user feature. That'll make our life and users' life simpler.

**Pull request closed without merging**